rfcs/058-relevance-testing/README.md

```diff
@@ -0,0 +1,147 @@
+
+Pytest's [parametrised tests](https://docs.pytest.org/en/7.1.x/example/parametrize.html) and [fixtures](https://docs.pytest.org/en/6.2.x/fixture.html) might help us achieve this plain-language test-writing style.
+
+### Rank eval API requests
+
+Our testing needs have developed over time, and we rarely use elasticsearch's rank eval API in the way that it's supposed to be used. In many cases, the requests we use to analyse quality are straightforward `search` requests.
+
+The outputs of our tests are often more binary than the rank eval API's responses, and we're discarding information about scoring which might be useful.
+In other cases, we've extended our code to test things which the API doesn't support, eg relative positions of expected results.
+
+These differences are hard to understand from the code, and are not well documented. A new implementation should be clearer about where those divergences are.
+
+### Test indices
+
+At the moment, we're able to test against indices copied from the catalogue API, ie `works` and `images`. We'd like to be able to test against other types of content like articles, exhibitions, events, functional content, and concepts.
+
+These source indices live in different clusters, so supporting cross cluster replication into rank from multiple clusters in future would be necessary.
+
+### Test outputs
+
+We know that optimising search relevance is a game of compromises, and that we're unlikely to be able to satisfy every search intention perfectly. In other words, we expect some of our tests to fail every time, even when search quality is good.
+
+For example, we might run a test for a new candidate mapping/query where document `abc` is expected to appear as the first result. A new version of the index might cause it to appear as the second result instead. The current version of rank would consider this a failure, and would alert us to the regression.
+
+To keep things moving in cases where we're satisfied with the overall search quality, users can currently set a `knownFailure` flag on individual tests, allowing the full suite to pass as a whole even when individual tests fail.
+
+`knownFailure`s are a useful but ugly bandage over an interesting problem. They obscure the severity of each failure during experiments, and make it harder to evaluate the development of search relevance over time.
+
+While the example above might represent a minor degradation in search quality for one intention, it's not something we would normally consider a catastrophic failure. If the same change led to an improvement in search quality for a different set of intentions, we might still want to deploy it.
+
+The scoring and passing of tests should be more nuanced, and should be able to account for the _ideal_ and _worst-case_ scenarios for each test.
+
+We should still be alerted to any degradations in individual or overall scores in CI, and we should be able to set extreme thresholds for each test which would cause rank to fail.
```

Comment:
What would this alerting look like? Does pytest have any ability to report on anything more than binary success/failure?

Reply:
I think it would be useful to show a summary of the scores at the end of each test run, even in cases where the tests pass without warnings.

It's possible to augment pytest's outputs with extra context by writing extra [plugins](https://docs.pytest.org/en/7.1.x/how-to/writing_plugins.html#writing-plugins). By defining a new `pytest_report_header()` or `pytest_terminal_summary()` etc we can be much more explicit about what the output of each test (or group of tests) means.

In those cases where scores are below an ideal threshold but aren't so bad that we want the whole suite to fail, we can also raise [warnings](https://docs.pytest.org/en/7.1.x/how-to/capture-warnings.html#warns), which fall into a separate section of the output.

Reply:
Pytest is impressively/concerningly configurable!